File: ios/StockbookTests/DaySummaryDocumentTests.swift
import Testing
import Foundation
@testable import Stockbook

/// The day, laid out for printing.
///
/// The twin of `DaySummaryDocumentTests.kt`, test for test. `DayBookTests` pins
/// the arithmetic; this pins what the owner actually reads — that a section is
/// only there when it has something in it, that a bill nobody paid for says so
/// on its own row, and that the foot of the page states the same three figures
/// the book computed rather than a fourth opinion about them.
@Suite("Day summary document")
@MainActor
struct DaySummaryDocumentTests {

    private let english = Strings(language: .english)

    private var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }

    private func at(_ hour: Int, day: Int = 22) -> Date {
        calendar.date(from: DateComponents(year: 2026, month: 8, day: day, hour: hour))!
    }

    private var day: Date { at(9) }

    private func makeStore() -> StockbookStore {
        let store = StockbookStore(repository: InMemoryRepository())
        store.setOwnerName("Al Salam Hardware")
        return store
    }

    private func page(_ store: StockbookStore) -> DaySummaryDocument {
        DaySummaryDocument.forDay(
            book: store.dayBook(day, calendar: calendar),
            settings: store.settings,
            strings: english
        )
    }

    @Test("The page says whose it is, what it is, and which day")
    func heading() {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: nil, amount: 100, createdAt: at(9))

        let document = page(store)

        #expect(document.shopName == "Al Salam Hardware")
        #expect(document.onDate == english.longDate(day))
        // The word matters. This names everybody who was billed and everything
        // the shop spent — it is not one party's account and must never carry
        // the word that means one.
        #expect(document.title == "Day Summary")
        #expect(!document.title.lowercased().contains("statement"))
    }

    @Test("Sections come in the order the day is read, and only where there is something")
    func sectionsPresent() {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: 0, amount: 100, createdAt: at(9))
        store.recordPayment(customerKey: Customer.key(for: "Ahmed"), amount: 50, receivedAt: at(10))
        store.addExpense(amount: 30, note: "Petrol", spentAt: at(11))

        // No deliveries, no supplier payments, no credit notes that day — and so
        // no headings for them. A heading over nothing is a question the reader
        // has to answer themselves.
        #expect(page(store).sections.map(\.heading) == ["Bills", "Received", "Expenses"])
    }

    @Test("A section totals what those things came to")
    func sectionSubtotal() throws {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: 0, amount: 100, createdAt: at(9))
        store.saveBill(customer: "Khalid", paid: nil, amount: 250.5, createdAt: at(10))

        let bills = try #require(page(store).sections.first)

        #expect(bills.subtotalLabel == "Subtotal")
        // What was sold, not what was collected for it. The cash foot is where
        // that question gets answered, once, for the whole day.
        #expect(bills.subtotalValue == "SAR 350.50")
    }

    @Test("A bill still owed for says how much, and one paid in full says nothing")
    func creditIsVisible() throws {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: 40, amount: 100, createdAt: at(9), invoiceNo: "6356")
        store.saveBill(customer: "Khalid", paid: nil, amount: 50, createdAt: at(10), invoiceNo: "6357")

        let rows = try #require(page(store).sections.first).rows

        #expect(rows[0].detail == "Invoice #6356 · SAR 60 on credit")
        #expect(rows[1].detail == "Invoice #6357")
    }

    @Test("A credit note never says on credit")
    func creditNoteDetail() throws {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: 0, amount: 500, createdAt: at(9))
        store.addCreditNote(customerKey: Customer.key(for: "Ahmed"), amount: 120, noteNo: "22", issuedAt: at(10))

        let note = try #require(page(store).sections.last?.rows.first)

        // It settles nothing by design. Marking the whole of it "on credit"
        // would be saying money is owed that was never going to be paid.
        #expect(note.detail == "Credit Note #22")
    }

    @Test("A bill lists what was on it, under its own row")
    func itemsUnderTheRow() throws {
        let store = makeStore()
        let padlock = store.addProduct(name: "Padlock 40mm", stock: 10, cost: 20, price: 30)
        store.saveBill(
            lines: [DraftLine(productUID: padlock.uid, qty: 3, price: 30)],
            customer: "Ahmed",
            paid: nil,
            createdAt: at(9)
        )

        let item = try #require(page(store).sections.first?.rows.first?.items.first)

        #expect(item.text == "3 × Padlock 40mm")
        #expect(item.amount == "SAR 90")
    }

    @Test("The foot states the day's cash, and the net may be negative")
    func cashFoot() throws {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: nil, amount: 100, createdAt: at(9))
        let supplier = try #require(store.addSupplier(name: "Gulf Locks"))
        store.recordSupplierBill(supplierKey: supplier.key, amount: 250, createdAt: at(10))

        let cash = page(store).cash

        #expect(cash.map(\.label) == ["Money in", "Money out", "Net for the day"])
        #expect(cash.map(\.value) == ["SAR 100", "SAR 250", "-SAR 150"])
        // The one figure the eye should stop on, and the only line marked.
        #expect(cash.map(\.isNet) == [false, false, true])
    }

    @Test("Money in is what the book says, never what the rows add up to")
    func footIsNotTheSubtotal() throws {
        let store = makeStore()
        // Three hundred sold, forty of it taken. A page whose foot read 300
        // would have the owner hunting for money nobody paid.
        store.saveBill(customer: "Ahmed", paid: 40, amount: 100, createdAt: at(9))
        store.saveBill(customer: "Khalid", paid: 0, amount: 200, createdAt: at(10))

        let document = page(store)

        #expect(try #require(document.sections.first).subtotalValue == "SAR 300")
        #expect(document.cash.first?.value == "SAR 40")
    }

    @Test("A day nothing happened on states that and nothing else")
    func emptyDay() {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: nil, amount: 100, createdAt: at(9, day: 20))

        let document = page(store)

        #expect(document.isEmpty)
        #expect(document.emptyLine == "Nothing was recorded on this day.")
        // No cash foot either: a page with no figures on it must not state a
        // cash position, even a zero one, for a day it knows nothing about.
        #expect(document.cash.isEmpty)
    }

    @Test("The paper is called what the statement calls it")
    func referencesMatchTheStatement() throws {
        let store = makeStore()
        store.saveBill(customer: "Ahmed", paid: nil, amount: 100, createdAt: at(9))
        store.recordPayment(customerKey: Customer.key(for: "Ahmed"), amount: 50, receivedAt: at(10), paymentNo: "1024")
        let supplier = try #require(store.addSupplier(name: "Gulf Locks"))
        store.recordSupplierBill(supplierKey: supplier.key, amount: 300, createdAt: at(11), invoiceNo: "88")
        store.addExpense(amount: 30, note: "Petrol", spentAt: at(12))

        let details = page(store).sections.flatMap(\.rows).map(\.detail)

        #expect(details == [
            // No paper number on the bill, so the app's own counter — the same
            // fallback the statement uses.
            "Bill #1",
            "Payment #1024",
            "Delivery #88",
            // Joined to nobody, numbered by nobody.
            nil
        ])
    }

    @Test("An expense is named by what it went on")
    func expenseRow() throws {
        let store = makeStore()
        store.addExpense(amount: 30, note: "Petrol", spentAt: at(9))

        let row = try #require(page(store).sections.first?.rows.first)

        #expect(row.name == "Petrol")
        #expect(row.amount == "SAR 30")
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/DaySummaryDocumentTests.kt
package com.stockbook.core

import com.stockbook.core.model.Customer
import com.stockbook.core.store.DraftLine
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.DaySummaryDocument
import com.stockbook.core.text.Strings
import java.time.Instant
import java.time.ZoneId
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * The day, laid out for printing.
 *
 * `DayBookTests` pins the arithmetic; this pins what the owner actually reads —
 * that a section is only there when it has something in it, that a bill nobody
 * paid for says so on its own row, and that the foot of the page states the same
 * three figures the book computed rather than a fourth opinion about them.
 */
class DaySummaryDocumentTests {

    private val strings = Strings(AppLanguage.ENGLISH)
    private val zone: ZoneId = ZoneId.of("UTC")
    private val day: Instant = Instant.parse("2026-08-22T09:00:00Z")

    private fun at(hour: Int): Instant = Instant.parse("2026-08-22T%02d:00:00Z".format(hour))

    private fun store() = StockbookStore(InMemoryRepository()).also { it.setOwnerName("Al Salam Hardware") }

    private fun StockbookStore.page() =
        DaySummaryDocument.forDay(dayBook(day, zone), settings, strings)

    @Test
    fun `the page says whose it is, what it is, and which day`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = null, amount = 100.0, createdAt = at(9))

        val page = store.page()

        assertEquals("Al Salam Hardware", page.shopName)
        assertEquals(strings.longDate(day), page.onDate)
        // The word matters. This names everybody who was billed and everything
        // the shop spent — it is not one party's account and must never carry
        // the word that means one.
        assertEquals("Day Summary", page.title)
        assertFalse(page.title.contains("Statement", ignoreCase = true))
    }

    @Test
    fun `sections come in the order the day is read, and only where there is something`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 100.0, createdAt = at(9))
        store.recordPayment(Customer.key("Ahmed"), 50.0, receivedAt = at(10))
        store.addExpense(30.0, "Petrol", spentAt = at(11))

        // No deliveries, no supplier payments, no credit notes that day — and so
        // no headings for them. A heading over nothing is a question the reader
        // has to answer themselves.
        assertEquals(listOf("Bills", "Received", "Expenses"), store.page().sections.map { it.heading })
    }

    @Test
    fun `a section totals what those things came to`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 100.0, createdAt = at(9))
        store.saveBill(customer = "Khalid", paid = null, amount = 250.5, createdAt = at(10))

        val bills = store.page().sections.single()

        assertEquals("Subtotal", bills.subtotalLabel)
        // What was sold, not what was collected for it. The cash foot is where
        // that question gets answered, once, for the whole day.
        assertEquals("SAR 350.50", bills.subtotalValue)
    }

    @Test
    fun `a bill still owed for says how much, and one paid in full says nothing`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = 40.0, amount = 100.0, invoiceNo = "6356", createdAt = at(9))
        store.saveBill(customer = "Khalid", paid = null, amount = 50.0, invoiceNo = "6357", createdAt = at(10))

        val rows = store.page().sections.single().rows

        assertEquals("Invoice #6356 · SAR 60 on credit", rows[0].detail)
        assertEquals("Invoice #6357", rows[1].detail)
    }

    @Test
    fun `a credit note never says on credit`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 500.0, createdAt = at(9))
        store.addCreditNote(Customer.key("Ahmed"), amount = 120.0, noteNo = "22", issuedAt = at(10))

        val note = store.page().sections.last().rows.single()

        // It settles nothing by design. Marking the whole of it "on credit"
        // would be saying money is owed that was never going to be paid.
        assertEquals("Credit Note #22", note.detail)
    }

    @Test
    fun `a bill lists what was on it, under its own row`() {
        val store = store()
        val padlock = store.addProduct("Padlock 40mm", stock = 10, cost = 20.0, price = 30.0)
        store.saveBill(
            lines = listOf(DraftLine(padlock.uid, qty = 3, price = 30.0)),
            customer = "Ahmed",
            paid = null,
            createdAt = at(9)
        )

        val item = assertNotNull(store.page().sections.single().rows.single().items.firstOrNull())

        assertEquals("3 × Padlock 40mm", item.text)
        assertEquals("SAR 90", item.amount)
    }

    @Test
    fun `the foot states the day's cash, and the net may be negative`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = null, amount = 100.0, createdAt = at(9))
        val supplier = assertNotNull(store.addSupplier("Gulf Locks"))
        store.recordSupplierBill(supplier.key, amount = 250.0, createdAt = at(10))

        val cash = store.page().cash

        assertEquals(listOf("Money in", "Money out", "Net for the day"), cash.map { it.label })
        assertEquals(listOf("SAR 100", "SAR 250", "-SAR 150"), cash.map { it.value })
        // The one figure the eye should stop on, and the only line marked.
        assertEquals(listOf(false, false, true), cash.map { it.isNet })
    }

    @Test
    fun `money in is what the book says, never what the rows add up to`() {
        val store = store()
        // Three hundred sold, forty of it taken. A page whose foot read 300
        // would have the owner hunting for money nobody paid.
        store.saveBill(customer = "Ahmed", paid = 40.0, amount = 100.0, createdAt = at(9))
        store.saveBill(customer = "Khalid", paid = 0.0, amount = 200.0, createdAt = at(10))

        val page = store.page()

        assertEquals("SAR 300", page.sections.single().subtotalValue)
        assertEquals("SAR 40", page.cash.first().value)
    }

    @Test
    fun `a day nothing happened on states that and nothing else`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = null, amount = 100.0, createdAt = Instant.parse("2026-08-20T09:00:00Z"))

        val page = store.page()

        assertTrue(page.isEmpty)
        assertEquals("Nothing was recorded on this day.", page.emptyLine)
        // No cash foot either: a page with no figures on it must not state a
        // cash position, even a zero one, for a day it knows nothing about.
        assertTrue(page.cash.isEmpty())
    }

    @Test
    fun `the paper is called what the statement calls it`() {
        val store = store()
        store.saveBill(customer = "Ahmed", paid = null, amount = 100.0, createdAt = at(9))
        store.recordPayment(Customer.key("Ahmed"), 50.0, paymentNo = "1024", receivedAt = at(10))
        val supplier = assertNotNull(store.addSupplier("Gulf Locks"))
        store.recordSupplierBill(supplier.key, amount = 300.0, invoiceNo = "88", createdAt = at(11))
        store.addExpense(30.0, "Petrol", spentAt = at(12))

        val details = store.page().sections.flatMap { it.rows }.map { it.detail }

        assertEquals(
            listOf(
                // No paper number on the bill, so the app's own counter — the
                // same fallback the statement uses.
                "Bill #1",
                "Payment #1024",
                "Purchase #88",
                // Joined to nobody, numbered by nobody.
                null
            ),
            details
        )
    }

    @Test
    fun `an expense is named by what it went on`() {
        val store = store()
        store.addExpense(30.0, "Petrol", spentAt = at(9))

        val row = store.page().sections.single().rows.single()

        assertEquals("Petrol", row.name)
        assertEquals("SAR 30", row.amount)
    }
}
